tests/attacks/test_knockoff_nets.py

```diff
@@ -0,0 +1,434 @@
+
+        # Logits layer
+        logits = tf.layers.dense(flattened, 10)
+
+        # Train operator
+        loss = tf.reduce_mean(tf.losses.softmax_cross_entropy(logits=logits, onehot_labels=output_ph))
+        optimizer = tf.train.AdamOptimizer(learning_rate=0.001)
+        train = optimizer.minimize(loss)
+
+        # TensorFlow session and initialization
+        sess.run(tf.global_variables_initializer())
+
+        # Create the classifier
+        thieved_tfc = TensorFlowClassifier(clip_values=(0, 1), input_ph=input_ph, output=logits, labels_ph=output_ph,
+                                           train=train, loss=loss, learning=None, sess=sess)
+
+        # Create random attack
+        attack = KnockoffNets(classifier=victim_tfc, batch_size_fit=BATCH_SIZE, batch_size_query=BATCH_SIZE,
+                              nb_epochs=NB_EPOCHS, nb_stolen=NB_STOLEN, sampling_strategy='random')
+        thieved_tfc = attack.extract(x=self.x_train, thieved_classifier=thieved_tfc)
+
+        victim_preds = np.argmax(victim_tfc.predict(x=self.x_train[:100]), axis=1)
+        thieved_preds = np.argmax(thieved_tfc.predict(x=self.x_train[:100]), axis=1)
+        acc = np.sum(victim_preds == thieved_preds) / len(victim_preds)
+
+        self.assertGreater(acc, 0.3)
+
+        # Create adaptive attack
+        attack = KnockoffNets(classifier=victim_tfc, batch_size_fit=BATCH_SIZE, batch_size_query=BATCH_SIZE,
+                              nb_epochs=NB_EPOCHS, nb_stolen=NB_STOLEN, sampling_strategy='adaptive', reward='all')
+        thieved_tfc = attack.extract(x=self.x_train, y=self.y_train, thieved_classifier=thieved_tfc)
+
+        victim_preds = np.argmax(victim_tfc.predict(x=self.x_train[:100]), axis=1)
+        thieved_preds = np.argmax(thieved_tfc.predict(x=self.x_train[:100]), axis=1)
+        acc = np.sum(victim_preds == thieved_preds) / len(victim_preds)
+
+        self.assertGreater(acc, 0.4)
+
+        # Clean-up session
+        sess.close()
+        tf.reset_default_graph()
+
+    @unittest.skipIf(tf.__version__[0] == '2', reason='Skip unittests for Tensorflow v2 until Keras supports Tensorflow'
+                                                      ' v2 as backend.')
+    def test_krclassifier(self):
+        """
+        Second test with the KerasClassifier.
+        :return:
+        """
+        # Build KerasClassifier
+        victim_krc = get_classifier_kr()
+
+        # Create simple CNN
+        model = Sequential()
+        model.add(Conv2D(1, kernel_size=(7, 7), activation='relu', input_shape=(28, 28, 1)))
+        model.add(MaxPooling2D(pool_size=(4, 4)))
+        model.add(Flatten())
+        model.add(Dense(10, activation='softmax'))
+        loss = keras.losses.categorical_crossentropy
+        model.compile(loss=loss, optimizer=keras.optimizers.Adam(lr=0.001), metrics=['accuracy'])
+
+        # Get classifier
+        thieved_krc = KerasClassifier(model, clip_values=(0, 1), use_logits=False)
```

Comment:
Would it be possible to create thieved_krc with get_classifier_tf?

Reply:
Solved.